feat: generate functions for all message types

ajcasagrande opened the pull request:
Signed-off-by: Anthony Casagrande <anthony.j.casagrande@intel.com>

> Note: I will leave comments below on many of the files to help make code review a little easier

There are a lot of commands defined in the code that are not accessible because they do not have proper mappings. This code consists of a python script that scrapes the response types and formulates functions based on that. It may not be perfect, but I do not believe there is any missing commands from before the generator ran.

Also, fixes a small bug in the `DevicesFromProbeResponses` where there is no http timeout. While not usually going to be an issue, better be safe than sorry. Was discovered when running the unit tests.

ajcasagrande commented on constant.go:
```diff
@@ -1,114 +1,5 @@
 package onvif
 
-// Onvif WebService
```
this whole deleted section is now auto-generated as `names.go`

ajcasagrande commented on analytics/function.go:
```diff
@@ -1,121 +1,115 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
```
this file is auto-generated

ajcasagrande commented on event/function.go:
```diff
@@ -1,61 +1,97 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
```
this file is auto-generated

ajcasagrande commented on device/function.go:
```diff
@@ -1,251 +1,817 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
```
this file is auto-generated

ajcasagrande commented on event/operation.go:
```diff
@@ -1,145 +0,0 @@
-package event
```
the code in this file was copied as-is to `types.go` to better match the rest of the packages, and to prevent having to write custom case for just this service.

ajcasagrande commented on event/types.go:
```diff
@@ -227,3 +227,143 @@ type NotifyMessageNotSupportedFault struct {
 //SubscribeCreationFailedFault response type
 type SubscribeCreationFailedFault struct {
 }
+
```
Moved from `event/operation.go`

ajcasagrande commented on functionmap.go:
```diff
@@ -2,166 +2,31 @@ package onvif
-	ModifyRules:       &analytics.ModifyRulesFunction{},
-}
-
 func FunctionByServiceAndFunctionName(serviceName, functionName string) (Function, error) {
```
I removed all of the mappings above, and they are now auto-generated in `mappings.go`. The only function left here is this one, and I have slimmed it down by reducing the duplicated code.

ajcasagrande commented on media2/function.go:
```diff
@@ -1,41 +1,43 @@
 }
-
-func (function *RemoveConfigurationFunction) Response() interface{} {
+func (_ *RemoveConfigurationFunction) Response() interface{} {
```
Note: `media2/type*.go` was renamed to `media2/types*.go` in order to be consistant with the other packages.

weichou1229 commented on python/gen_commands.py:
```diff
@@ -0,0 +1,138 @@
+            for generator in self.generators:
+                generator.write_func_map(w)
+
+    def _generate_constants(self):
```
rename to  _generate_names?

ajcasagrande replied:
Ah ok I see. The reason I called it constants was because it was generating the `const` values, but the file was actually named `names.go`, so yeah `_generate_names` probably makes more sense.

weichou1229 commented:
> This code consists of a python script that scrapes the response types and formulates functions based on that. 

So If the user wants to add a new ONVIF function,
- user needs to add the definition in types.go file 
- and run the gen_commands.py to generate the mapping and function code for the device-onvif-camera. 

Could you help add this new feature to the doc?

ajcasagrande commented:
> > This code consists of a python script that scrapes the response types and formulates functions based on that.
> 
> So If the user wants to add a new ONVIF function,
> 
> * user needs to add the definition in types.go file
> * and run the gen_commands.py to generate the mapping and function code for the device-onvif-camera.
> 
> @ajcasagrande Could you help resolve the conflicts and add this new feature to the doc?

@weichou1229 Yes, you are correct, but some clarification: For the second bullet, this only allows the consumer of this library (such as device-onvif-camera) to make those calls from go code, however it will not expose them to EdgeX command service, as they need to be added to the device profile for that.

I can add some documentation for it, and the conflicts should be easy to resolve.

ajcasagrande commented:
@weichou1229 Ready for re-review. I have:

- Added documentation
- Added `//go:generate` lines as well as `DO NOT EDIT` lines
- Added empty response types for `Imaging` web service and function mapping support

Note: This PR should not be used in `device-onvif-camera` until post-release just to be safe.

weichou1229 reviewed (approved):
LGTM.
